### PR TITLE
search: only give timeout alert if we had no results

### PR DESCRIPTION
**Pull request by keegancsmith**
This PR removes the need for resultsWithTimeoutSuggestion to exist. resultsWithTimeoutSuggestion was added so other code could call doResults and get back a context.DeadlineExceeded if we timedout. However, the code has evolved a lot since then. Currently the only way doResults returns DeadlineExceeded is if determineRepos times out (rare). This commit makes it such that doResults never returns DeadlineExceeded and consistently sets the timeout alert iff we had no results and all repos timed out.

To do this we made determineRepos check for a timeout. Additionally we had to track the start time for the alert. This is now done in a more correct place, SearchInputs.

This was tested by doing lots of manual testing. In particular around AND queries as well as searching across lots of unindexed branches.

Fixes https://github.com/sourcegraph/sourcegraph/issues/20319

### CODE REVIEW

**Comment by keegancsmith** [comment]
Manual testing seems to indicate that the first commit may be incorrect. When I test a query like

```
repo:^github\.com/sourcegraph/sourcegraph$ rev:*refs/heads/* timeout:1s (doResults and resultsWithTimeout)
```

it does not give the timeout alert. So will spend a bit more time on this.

File: cmd/frontend/graphqlbackend/search_alert.go [hunk 1]
```diff
@@ -750,6 +753,10 @@ func (o *alertObserver) Done(stats *streaming.Stats) (*searchAlert, error) {
 		o.update(alertForStructuralSearchNotSet(o.Inputs.OriginalQuery))
 	}
 
+	if !o.hasResults && stats.Status.All(search.RepoStatusTimedout) && stats.Status.Len() == len(stats.Repos) {
+		return alertForTimeout(o.Inputs), nil
```
Comment (camdencheek) [inline cmd/frontend/graphqlbackend/search_alert.go:757]: Will "timed out" always be the highest priority alert? This is currently overriding any other alerts. Maybe just update instead, and make `alertForTimeout` high priority?

```suggestion
		o.update(alertForTimeout(o.Inputs))
```

**Review by camdencheek — APPROVED** [review]

File: cmd/frontend/graphqlbackend/search.go [hunk 2]
```diff
@@ -137,6 +140,7 @@ func NewSearchImplementer(ctx context.Context, db dbutil.DB, args *SearchArgs) (
 			Pagination:     pagination,
 			PatternType:    searchType,
 			DefaultLimit:   defaultLimit,
+			Start:          start,
```
Comment (stefanhengl) [inline cmd/frontend/graphqlbackend/search.go:143]: Is there a way to keep `SearchInputs` immutable without making things more complicated?

**Review by stefanhengl — APPROVED** [review]

File: cmd/frontend/graphqlbackend/search.go (hunk 2 above)
Comment (keegancsmith) [reply cmd/frontend/graphqlbackend/search.go:143]: It is immutable, we only set this once :)

Maybe you want it to be deterministic/pure function. That seems reasonable, it's just so convenient to track this here. To be honest, we had an incorrect start values in a few places in our backend so this feels good. But thinking more about it, this start value is used to work out the timeout. We can just ask searchinputs what the timeout is now! Will update the PR :)

Comment (stefanhengl) [reply cmd/frontend/graphqlbackend/search.go:143]: > It is immutable, we only set this once :)

yeah, terrible wording on my side :-), deterministic is what I meant.

**Review by rvantonder — APPROVED** [review]
Rubber stamping.

**Comment by lguychard** [comment]
@keegancsmith what's the status of this PR?

**Review by rmmh — APPROVED** [review]

**Comment by keegancsmith** [comment]
> @keegancsmith what's the status of this PR?

sorry in my personal backlog. Will try get it out before end of week.

**Comment by keegancsmith** [comment]
Ok did some more testing after rebasing. Everything is working. Will address the inline comments above then ship this!

**Comment by camdencheek** [comment]
Oops. I thought this was already merged when I merged #21162. Will likely result in some conflicts again -- sorry!

**Comment by keegancsmith** [comment]
@camdencheek no worries, I need to get this merged in lol. Totally my fault for having this on my backlog.

**Comment by keegancsmith** [comment]
I maybe did a bit too much work here and included a bit of refactoring. So @rvantonder and @camdencheek mind taking a close look at https://github.com/sourcegraph/sourcegraph/pull/20356/commits/19d27534dfcc6bbd3d46786cfd0df949458f3eb1 ?

**Comment by camdencheek** [comment]
I will never complain about reviewing refactors :)

**Review by camdencheek — APPROVED** [review]
Looks great! Great simplification -- I've stumbled over a lot of this complexity during my refactors, so I'm glad to see it gone

**Comment by keegancsmith** [comment]
The PR that will never merge. Legit integration test failures:
```
--- FAIL: TestSearch/graphql (3.81s)
--
  | --- FAIL: TestSearch/graphql/timeout_search_options (0.01s)
  | search_test.go:551: request GraphQL: context deadline exceeded
  | --- FAIL: TestSearch/stream (14.90s)
  | --- FAIL: TestSearch/stream/timeout_search_options (0.01s)
  | search_test.go:555: Want search alert but got nil
```

**Comment by keegancsmith** [comment]
This PR is very out of date. More appropriate to tackle this with a fresh set of eyes.

**Pull request closed without merging**